internal/telemetry/client.go

```diff
@@ -0,0 +1,407 @@
+			return v
+		}
+		return def
+	}
+	c.Service = fromEnvOrDefault("DD_SERVICE", c.Service)
+	if len(c.Service) == 0 {
+		// I think service *has* to be something?
+		c.Service = "unnamed-go-service"
+	}
+	c.Env = fromEnvOrDefault("DD_ENV", c.Env)
+	c.Version = fromEnvOrDefault("DD_VERSION", c.Version)
+
+	c.APIKey = fromEnvOrDefault("DD_API_KEY", c.APIKey)
+	// TODO: Initialize URL/endpoint from environment var
+
+	r := c.newRequest(RequestTypeAppStarted)
+	r.Payload = payload
+	c.schedulesubmit(r)
+	c.flush()
+
+	if c.SubmissionInterval == 0 {
+		c.SubmissionInterval = 60 * time.Second
+	}
+	c.t = time.AfterFunc(c.SubmissionInterval, c.backgroundFlush)
+}
+
+// Stop notifies the telemetry endpoint that the app is closing. All outstanding
+// messages will also be sent. No further messages will be sent until the client
+// is started again
+func (c *Client) Stop() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if !c.started {
+		return
+	}
+	defer func() { c.started = false }()
+	c.t.Stop()
+	// close request types have no body
+	r := c.newRequest(RequestTypeAppClosing)
+	c.schedulesubmit(r)
+	c.flush()
+}
+
+type metricKind string
+
+var (
+	metricKindGauge metricKind = "gauge"
+	metricKindCount metricKind = "count"
+)
+
+type metric struct {
+	name  string
+	kind  metricKind
+	value float64
+	// Unix timestamp
+	ts     float64
+	tags   []string
+	common bool
+}
+
+// TODO: Can there be identically named/tagged metrics with a "common" and "not
+// common" variant?
+
+func newmetric(name string, kind metricKind, tags []string, common bool) *metric {
+	return &metric{
+		name:   name,
+		kind:   kind,
+		tags:   append([]string(nil), tags...),
+		common: common,
+	}
+}
+
+func metricKey(name string, tags []string) string {
+	return name + strings.Join(tags, "-")
+}
+
+// Gauge sets the value for a gauge with the given name and tags. If the metric
+// is not language-specific, common should be set to true
+func (c *Client) Gauge(name string, value float64, tags []string, common bool) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if !c.started {
+		return
+	}
+	key := metricKey(name, tags)
+	m, ok := c.metrics[key]
+	if !ok {
+		m = newmetric(name, metricKindGauge, tags, common)
+		c.metrics[key] = m
+	}
+	m.value = value
+	m.ts = float64(time.Now().Unix())
+	c.newMetrics = true
+}
+
+// Count adds the value to a count with the given name and tags. If the metric
+// is not language-specific, common should be set to true
+func (c *Client) Count(name string, value float64, tags []string, common bool) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if !c.started {
+		return
+	}
+	key := metricKey(name, tags)
+	m, ok := c.metrics[key]
+	if !ok {
+		m = newmetric(name, metricKindCount, tags, common)
+		c.metrics[key] = m
+	}
+	m.value += value
+	m.ts = float64(time.Now().Unix())
+	c.newMetrics = true
+}
+
+// Flush sends any outstanding telemetry messages and aggregated metrics to be
+// sent to the backend. Requests are sent in the background
+func (c *Client) Flush() {
```

Comment:
Q: Given that `Stop()` also flushes, what's the use case for exposing this as a public method?

Reply:
Good point, I'll remove it 👍